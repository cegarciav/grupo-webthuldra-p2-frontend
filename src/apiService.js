import axios from 'axios';

const apiOrigin = process.env.REACT_APP_API_ORIGIN;

const apiRequester = axios.create({
  baseURL: apiOrigin,
});

const errorGenerator = (e) => {
  let errorResponse = null;
  if (e.response) {
    errorResponse = {
      errors: e.response.data.errors,
      type: 'response',
      statusCode: e.response.status,
    };
  } else {
    errorResponse = {
      type: 'unexpected',
      message: 'Algo salió mal. Por favor, revisa tu conexión o intenta ingresar más tarde',
    };
  }
  return errorResponse;
};

async function apiGet(path, params) {
  try {
    if (!apiOrigin) throw Error('La conexión con el servidor no ha podido ser establecida');
    const url = new URL(`${apiOrigin}${path}`);
    const token = localStorage.getItem('apiToken');
    let headers = {};
    if (token) headers = { Authorization: `bearer ${token}` };
    if (params) {
      Object.keys(params)
        .forEach((key) => url.searchParams.append(key, params[key]));
    }
    const response = await apiRequester.get(
      url, {
        headers,
      },
    );
    return {
      data: response.data,
      statusCode: response.status,
    };
  } catch (e) {
    return errorGenerator(e);
  }
}

async function apiPost(path, body, params) {
  try {
    if (!apiOrigin) throw Error('La conexión con el servidor no ha podido ser establecida');
    const url = new URL(`${apiOrigin}${path}`);
    const token = localStorage.getItem('apiToken');
    let headers = {};
    if (token) headers = { Authorization: `bearer ${token}` };
    if (params) {
      Object.keys(params)
        .forEach((key) => url.searchParams.append(key, params[key]));
    }
    const response = await apiRequester.post(
      url,
      body, {
        headers,
      },
    );
    return {
      data: response.data,
      statusCode: response.status,
    };
  } catch (e) {
    return errorGenerator(e);
  }
}

async function apiPatch(path, body, params) {
  try {
    if (!apiOrigin) throw Error('La conexión con el servidor no ha podido ser establecida');
    const url = new URL(`${apiOrigin}${path}`);
    const token = localStorage.getItem('apiToken');
    let headers = {};
    if (token) headers = { Authorization: `bearer ${token}` };
    if (params) {
      Object.keys(params)
        .forEach((key) => url.searchParams.append(key, params[key]));
    }
    const response = await apiRequester.patch(
      url,
      body, {
        headers,
      },
    );
    return {
      data: response.data,
      statusCode: response.status,
    };
  } catch (e) {
    return errorGenerator(e);
  }
}

async function apiDelete(path, body, params) {
  try {
    if (!apiOrigin) throw Error('La conexión con el servidor no ha podido ser establecida');
    const url = new URL(`${apiOrigin}${path}`);
    const token = localStorage.getItem('apiToken');
    let headers = {};
    if (token) headers = { Authorization: `bearer ${token}` };
    if (params) {
      Object.keys(params)
        .forEach((key) => url.searchParams.append(key, params[key]));
    }
    const response = await apiRequester.delete(
      url,
      body, {
        headers,
      },
    );
    return {
      data: response.data,
      statusCode: response.status,
    };
  } catch (e) {
    return errorGenerator(e);
  }
}

export {
  apiGet,
  apiPost,
  apiPatch,
  apiDelete,
};
