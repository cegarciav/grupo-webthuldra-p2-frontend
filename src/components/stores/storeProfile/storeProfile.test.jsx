import React from 'react';
import PropTypes from 'prop-types';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { cleanup } from '@testing-library/react';
import MainRouter from '../../../routers/index';
import { apiGet } from '../../../apiService';

function TestRouter({ path }) {
  return (
    <MemoryRouter initialEntries={[path]}>
      <MainRouter />
    </MemoryRouter>
  );
}

TestRouter.propTypes = {
  path: PropTypes.string.isRequired,
};

jest.mock('../../../apiService');

const storeId = '6f4e5985-2ff8-4118-bbaf-e6004d757c13';

const storeResponse = {
  storeId,
  address: 'Address 1',
  name: 'Name 1',
  description: 'Description 1',
  ownerId: '7d13c742-cd2d-45b7-9866-20dbb0eaa2e9',
  picture: null,
  owner: {
    firstName: 'Owner Name',
    lastName: 'Owner Lastname',
    email: 'owner@email.com',
  },
};

const commentsResponse = [
  {
    id: '26c08e2d-3caf-458b-9767-8477683668c8',
    text: 'Possitive comment',
    grade: 4,
    reviewerId: '96ab7b63-915e-442b-b83f-e0962bd2c8d8',
    storeId,
    store: {
      name: 'Name 1',
      address: 'Address 1',
    },
    reviewer: {
      firstName: 'Reviewer Name',
      lastName: 'Reviewer Lastname',
      email: 'reviewer@email.com',
    },
  },
  {
    id: '26c08e2d-3caf-458b-9767-8477683668c9',
    text: 'Negative comment',
    grade: 1,
    reviewerId: '96ab7b63-915e-442b-b83f-e0962bd2c8d8',
    storeId,
    store: {
      name: 'Name 1',
      address: 'Address 1',
    },
    reviewer: {
      firstName: 'Reviewer Name 2',
      lastName: 'Reviewer Lastname 2',
      email: 'reviewer2@email.com',
    },
  },
];

const productsResponse = [
  {
    id: 'a9aa09e7-83f8-4e95-b551-273cedf0afe3',
    name: 'Product 1',
    stock: 20,
    price: 500,
    unit: 'Unidad',
    storeId,
    picture: null,
  },
  {
    id: 'a3de523e-a264-4479-b1e9-d1da5d228caf',
    name: 'Product 2',
    stock: 15,
    price: 750,
    unit: 'Unidad',
    storeId,
    picture: null,
  },
  {
    id: '836fa52e-cece-4a60-ac9d-dabc46e50814',
    name: 'Product 3',
    stock: 10,
    price: 850,
    storeId,
    picture: null,
  },
];

describe('StoreProfile', () => {
  describe('When user is not logged-in', () => {
    it('renders the profile of the store', async () => {
      let storesRenderer;
      apiGet.mockImplementation((url) => {
        if (url === `/stores/${storeId}`) {
          return {
            data: storeResponse,
            statusCode: 200,
          };
        }
        if (url === `/stores/${storeId}/comments`) {
          return {
            data: commentsResponse,
            statusCode: 200,
          };
        }
        if (url === `/stores/${storeId}/products`) {
          return {
            data: productsResponse,
            statusCode: 200,
          };
        }
        return {};
      });
      await renderer.act(async () => {
        storesRenderer = await renderer.create(<TestRouter path={`/stores/${storeId}`} />);
      });
      const tree = storesRenderer.toJSON();
      expect(tree).toMatchSnapshot();
    });

    afterEach(() => {
      cleanup();
      jest.resetAllMocks();
    });
  });
});
