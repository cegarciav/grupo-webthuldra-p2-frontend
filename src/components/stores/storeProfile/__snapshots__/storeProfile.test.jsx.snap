// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoreProfile When user is not logged-in renders the profile of the store 1`] = `
<main>
  <section
    className="store-main-info"
  >
    <figure>
      <img
        alt="Description 1"
        src="/default_store_picture.png"
      />
    </figure>
    <main>
      <h1>
        Name 1
      </h1>
      <h4>
        Description 1
      </h4>
    </main>
  </section>
  <hr
    className="section-divisor"
  />
  <button
    type="button"
  >
    Editar tienda
  </button>
  <button
    type="button"
  >
    Eliminar tienda
  </button>
  <section
    className="store-main-profile"
  >
    <aside
      className="store-info"
    >
      <a
        className="user-summary"
        href="/users/7d13c742-cd2d-45b7-9866-20dbb0eaa2e9"
        onClick={[Function]}
      >
        <article>
          <h1>
            Owner Name Owner Lastname
          </h1>
          <section>
            <img
              alt="Owner Name Owner Lastname"
              src="/default_profile_picture.png"
            />
          </section>
          <p>
            Contacto: owner@email.com
          </p>
        </article>
      </a>
      <main
        className="comments-container"
      >
        <h4>
          Comentarios
        </h4>
        <ul
          className="comments-list"
        >
          <li>
            <article
              className="comment-summary"
            >
              <section
                className="user-info"
              >
                <a
                  href="/users/96ab7b63-915e-442b-b83f-e0962bd2c8d8"
                  onClick={[Function]}
                >
                  <figure>
                    <img
                      alt="Reviewer Name Reviewer Name"
                      src="/default_profile_picture.png"
                    />
                  </figure>
                  <p>
                    4/5
                  </p>
                </a>
              </section>
              <section
                className="comment-text"
              >
                Possitive comment
              </section>
            </article>
          </li>
          <li>
            <article
              className="comment-summary"
            >
              <section
                className="user-info"
              >
                <a
                  href="/users/96ab7b63-915e-442b-b83f-e0962bd2c8d8"
                  onClick={[Function]}
                >
                  <figure>
                    <img
                      alt="Reviewer Name 2 Reviewer Name 2"
                      src="/default_profile_picture.png"
                    />
                  </figure>
                  <p>
                    1/5
                  </p>
                </a>
              </section>
              <section
                className="comment-text"
              >
                Negative comment
              </section>
            </article>
          </li>
        </ul>
        <a
          href="/"
          role="button"
        >
          Escribir un comentario
        </a>
      </main>
    </aside>
    <section
      className="products-in-store"
    >
      <button
        type="button"
      >
        AÃ±adir producto
      </button>
      <button
        type="button"
      >
        Eliminar producto
      </button>
      <main>
        <h1>
          Lista de productos disponibles
        </h1>
        <ul
          className="products-list"
        >
          <li>
            <a
              className="product-summary"
              href="/stores/6f4e5985-2ff8-4118-bbaf-e6004d757c13/products/a9aa09e7-83f8-4e95-b551-273cedf0afe3"
              onClick={[Function]}
            >
              <article>
                <h1>
                  Product 1
                </h1>
                <section>
                  <img
                    alt="Product 1"
                    src="/default_product_picture.png"
                  />
                </section>
                <p>
                  Precio: $500 Unidad
                </p>
              </article>
            </a>
          </li>
          <li>
            <a
              className="product-summary"
              href="/stores/6f4e5985-2ff8-4118-bbaf-e6004d757c13/products/a3de523e-a264-4479-b1e9-d1da5d228caf"
              onClick={[Function]}
            >
              <article>
                <h1>
                  Product 2
                </h1>
                <section>
                  <img
                    alt="Product 2"
                    src="/default_product_picture.png"
                  />
                </section>
                <p>
                  Precio: $750 Unidad
                </p>
              </article>
            </a>
          </li>
          <li>
            <a
              className="product-summary"
              href="/stores/6f4e5985-2ff8-4118-bbaf-e6004d757c13/products/836fa52e-cece-4a60-ac9d-dabc46e50814"
              onClick={[Function]}
            >
              <article>
                <h1>
                  Product 3
                </h1>
                <section>
                  <img
                    alt="Product 3"
                    src="/default_product_picture.png"
                  />
                </section>
                <p>
                  Precio: $850 c/u
                </p>
              </article>
            </a>
          </li>
        </ul>
      </main>
    </section>
  </section>
</main>
`;
